File: javascript/148.Sort-List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
const getMid = (head) => {
  let slow = head;
  let fast = head.next;

  while (fast && fast.next) {
    slow = slow.next;
    fast = fast.next.next;
  }

  return slow;
};

const merge = (list1, list2) => {
  const dummy = new ListNode();
  let ptr = dummy;

  while (list1 && list2) {
    if (list1.val < list2.val) {
      ptr.next = list1;
      list1 = list1.next;
    } else {
      ptr.next = list2;
      list2 = list2.next;
    }

    ptr = ptr.next;
  }

  ptr.next = list1 ? list1 : list2;

  return dummy.next;
};

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var sortList = function (head) {
  if (!head || !head.next) {
    return head;
  }

  let left = head;
  const mid = getMid(head);
  const rightHead = mid.next;
  mid.next = null;
  let right = rightHead;

  left = sortList(left);
  right = sortList(right);
  return merge(left, right);
};

File: python/148.Sort-List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def getMid(self, node):
        slow = node
        fast = node

        while fast and fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next

        return slow

    def merge(self, l1, l2):
        dummy = ListNode()
        prev = dummy

        while (l1 and l2):
            if l1.val < l2.val:
                prev.next = l1
                l1 = l1.next
            else:
                prev.next = l2
                l2 = l2.next
            
            prev = prev.next

            prev.next = l1 or l2

        return dummy.next

    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        mid = self.getMid(head)
        rightHead = mid.next
        mid.next = None

        leftList = self.sortList(head)
        rightList = self.sortList(rightHead)

        return self.merge(leftList, rightList)